eclass/zig.eclass

```diff
@@ -0,0 +1,582 @@
+
+	ZBS_DEPENDENCIES_SRC_URI=
+
+	for dependency in "${!ZBS_DEPENDENCIES[@]}"; do
+		local uri="${ZBS_DEPENDENCIES[${dependency}]}"
+		if [[ "${#uri}" -gt 0 ]]; then
+			ZBS_DEPENDENCIES_SRC_URI+=" ${uri} -> ${dependency}"
+		else
+			# Unrepresentable dependency
+			continue
+		fi
+	done
+}
+_zig_set_zbs_uris
+
+# @ECLASS_VARIABLE: my_zbs_args
+# @DESCRIPTION:
+# Bash array with ebuild-specified arguments to pass to the
+# "zig build" after "src_configure".
+# It's appended to the ZBS_ARGS during "src_configure".  Note: if you
+# need to override default optimize mode of this eclass (ReleaseSafe)
+# with your default, please use "--release=small" etc. syntax so that
+# user can still override it in ZBS_ARGS_EXTRA.
+#
+# Example:
+# @CODE
+# src_configure() {
+# 	local my_zbs_args=(
+# 		-Dpie=true
+# 	)
+#
+# 	zig_src_configure
+# }
+# @CODE
+: "${my_zbs_args:=}"
+
+# @ECLASS_VARIABLE: ZBS_ARGS_EXTRA
+# @USER_VARIABLE
+# @DESCRIPTION:
+# Bash string with user-specified arguments to pass to the "zig build"
+# after "src_configure".
+# It's appended to the ZBS_ARGS during "zig_src_configure".
+#
+# If this does not have amount of jobs, eclass will try to take amount
+# of jobs from MAKEOPTS, and if it also does not have them, it will
+# default to $(nproc).
+#
+# Example:
+# @CODE
+# ZBS_ARGS_EXTRA="-j8 --release=small"
+# @CODE
+: "${ZBS_ARGS_EXTRA:=}"
+
+# @ECLASS_VARIABLE: ZBS_VERBOSE
+# @USER_VARIABLE
+# @DESCRIPTION:
+# If enabled, eclass will add "--summary all --verbose" options to
+# "ezig build", so that it prints every command before executing,
+# and summarry tree at the end of step.  If not, will do nothing.
+# Enabled by default.  Set to OFF to disable these verbose messages.
+#
+# Note: this variable does not control other options starting with
+# "--verbose-", such as "--verbose-link" or "--verbose-cimport".  If
+# you need them, add them manually to ZBS_ARGS_EXTRA.
+: "${ZBS_VERBOSE:=ON}"
+
+# @ECLASS_VARIABLE: BUILD_DIR
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Directory where all "ezig build" calls will be proceeded.
+# Defaults to "${WORKDIR}/${P}-build" if not set.
+: "${BUILD_DIR:=${WORKDIR}/${P}-build}"
+
+# @ECLASS_VARIABLE: ZBS_ECLASS_DIR
+# @DESCRIPTION:
+# Directory where various files used by this eclass are stored.
+# They can be supplied by the ebuild or by eclass.
+# Currently, it's used only for Zig packages, which are stored in "p/"
+# subdirectory.
+# Defaults to "${WORKDIR}/zig-eclass" if not set.
+# Should be set before calling "zig_src_unpack" or
+# "zig_live_fetch".
+: "${ZBS_ECLASS_DIR:=${WORKDIR}/zig-eclass}"
+
+# @FUNCTION: zig_get_jobs
+# @DESCRIPTION:
+# Returns number of jobs from ZBS_ARGS_EXTRA or MAKEOPTS.
+# If there is none, defaults to number of available processing units.
+zig_get_jobs() {
+	local all_args="${ZBS_ARGS_EXTRA} ${MAKEOPTS}"
+	local default_jobs="$(get_nproc)"
+	local jobs="$(makeopts_jobs "${all_args}" "${default_jobs}")"
```

Comment:
I suppose you should append them to one of the standard variables then. Also, I think they have lower priority here.

Reply:
Priority of "Zig over Make args" was fixed later, and if I use `get_makeopts_jobs` like this:

```bash
    local all_args="${MAKEOPTS} ${ZBS_ARGS_EXTRA}"
    local jobs="$(MAKEOPTS="${all_args}" get_makeopts_jobs -1)"

    if [[ "${jobs}" == "-1" || "${jobs}" == "0" ]]; then
        # Either variable was not found, or "-j0" was passed,
        # which is unsupported by Zig build system (no support
        # for unlimited parallelism like in make). Pass our
        # default number of jobs here.
        echo "$(get_nproc)"
    else
        echo "${jobs}"
    fi
```

It can be overwritten by 2 other variables and logic in this eclass starts to depend on order of other eclass, and just takes too much control in general:

```bash
_get_all_makeopts() {
    echo "${MAKEOPTS} ${GNUMAKEFLAGS} ${MAKEFLAGS}"
}
```

Reply:
CC @Flowdalic

Reply:
There does not seem much that can be done right now. The code snippet shown by @BratishkaErik could be improved a bit by replacing
```bash
local all_args="${MAKEOPTS} ${ZBS_ARGS_EXTRA}"
local jobs="$(MAKEOPTS="${all_args}" get_makeopts_jobs -1)"
```
with
```bash
local my_makeopts="${MAKEOPTS} ${GNUMAKEFLAGS} ${MAKEFLAGS} ${ZBS_ARGS_EXTRA}"
local jobs=$(makeopts_jobs ${my_makeopts} -1)
```

Other than that, modifications to `multiprocessing.eclass` are probably required. For example, instead of
```bash
get_makeopts_jobs() {
	makeopts_jobs "$(_get_all_makeopts)" "${1:-999}"
}
```
we could simply introduce an extra variable
```bash
get_makeopts_jobs() {
	makeopts_jobs "${ADDITIONAL_MAKEOPTS} $(_get_all_makeopts)" "${1:-999}"
}
```
then the above snippedcould be reduced to
```bash
    local all_args="${MAKEOPTS} ${ZBS_ARGS_EXTRA}"
    local jobs="$(ADDITIONAL_MAKEOPTS="${ZBS_ARGS_EXTRA}" get_makeopts_jobs -1)"

    if [[ "${jobs}" == "-1" || "${jobs}" == "0" ]]; then
        # Either variable was not found, or "-j0" was passed,
        # which is unsupported by Zig build system (no support
        # for unlimited parallelism like in make). Pass our
        # default number of jobs here.
        echo "$(get_nproc)"
    else
        echo "${jobs}"
    fi
```

But this would probably require a discussion, as there are many design approaches one could use to inject extra makeopts into `get_makeopts_(jobs|loadavg)`.